mirrord/agent/src/metrics.rs

```diff
@@ -0,0 +1,177 @@
+        "amount of unfiltered steal port subscriptions in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static STEAL_FILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
+    LazyLock::new(|| {
+        register_int_gauge!(
+            "mirrord_agent_steal_connection_subscription_count",
+            "amount of filtered connections in steal mode in mirrord-agent"
+        )
+        .expect("Valid at initialization!")
+    });
+
+pub(crate) static STEAL_UNFILTERED_CONNECTION_SUBSCRIPTION: LazyLock<IntGauge> =
+    LazyLock::new(|| {
+        register_int_gauge!(
+            "mirrord_agent_steal_connection_subscription_count",
+            "amount of unfiltered connections in steal mode in mirrord-agent"
+        )
+        .expect("Valid at initialization!")
+    });
+
+pub(crate) static TCP_OUTGOING_CONNECTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_tcp_outgoing_connection_count",
+        "amount of tcp outgoing connections in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+pub(crate) static UDP_OUTGOING_CONNECTION: LazyLock<IntGauge> = LazyLock::new(|| {
+    register_int_gauge!(
+        "mirrord_agent_udp_outgoing_connection_count",
+        "amount of udp outgoing connections in mirrord-agent"
+    )
+    .expect("Valid at initialization!")
+});
+
+#[derive(Error, Debug)]
+pub(crate) enum MetricsError {
+    #[error(transparent)]
+    FromUtf8(#[from] std::string::FromUtf8Error),
+
+    #[error(transparent)]
+    Prometheus(#[from] prometheus::Error),
+}
+
+impl IntoResponse for MetricsError {
+    fn into_response(self) -> axum::response::Response {
+        (http::StatusCode::INTERNAL_SERVER_ERROR, self.to_string()).into_response()
+    }
+}
+
+#[tracing::instrument(level = Level::TRACE,  ret, err)]
+async fn get_metrics() -> Result<String, MetricsError> {
+    use prometheus::{Encoder, TextEncoder};
+
+    let metric_families = prometheus::gather();
+
+    let mut buffer = Vec::new();
+    TextEncoder
+        .encode(&metric_families, &mut buffer)
+        .inspect_err(|error| tracing::error!(%error, "unable to encode prometheus metrics"))?;
+
+    Ok(String::from_utf8(buffer)?)
+}
+
+#[tracing::instrument(level = Level::TRACE, skip_all, ret ,err)]
+pub(crate) async fn start_metrics(
+    address: SocketAddr,
+    cancellation_token: CancellationToken,
+) -> Result<(), axum::BoxError> {
+    let app = Router::new().route("/metrics", get(get_metrics));
+
+    let listener = TcpListener::bind(address)
+        .await
+        .map_err(AgentError::from)
+        .inspect_err(|fail| tracing::error!(?fail, "Actor listener!"))?;
+
+    axum::serve(listener, app)
+        .with_graceful_shutdown(async move { cancellation_token.cancelled().await })
+        .await
+        .inspect_err(|fail| {
+            tracing::error!(%fail, "Could not start agent metrics
+        server!")
```

Comment:
I think you should cancel the token on error as well. Otherwise we'll have an agent running without metrics

Reply:
In general, a failure in this function should terminate the agent (and cancelling the token should terminate function)